File: scripts/c++/cinematic_model_v2.cpp
// Implementación de un robot por cables para el control
// de un efector final en diversas tareas.

#include <iostream>
#include <cmath>
#include <vector>
#include <map>
#include "matplotlibcpp.h"

namespace plt = matplotlibcpp;

// COMPILACIÓN, ENLAZADO Y EJECUCIÓN
// g++ cinematic_model_v2.cpp -o cinematic_model_v2 -I /usr/include/python3.12 -I $(python3 -c "import numpy; print(numpy.get_include())") -lpython3.12
// ./cinematic_model_v2

// VERIFICACIÓN DE LÍMITES
std::string 
verificar_limites(double x, double y, double largo_plano, double alto_plano, double largo_efector, double alto_efector)
{
    std::string error_msg = "";
    if (x >= (largo_plano - (largo_efector / 2)) || x <= (largo_efector / 2)) {
        error_msg += "\nFUERA DEL LÍMITE en x = " + std::to_string(x);
    }
    if (y >= (alto_plano - (alto_efector / 2)) || y <= (alto_efector / 2)) {
        error_msg += "\nFUERA DEL LÍMITE en y = " + std::to_string(y);
    }
    return error_msg;
}

// EFECTOR FINAL
void 
calcular_esquinas(double x, double y, double largo_efector, double alto_efector,
                  double& x1, double& y1, double& x2, double& y2,
                  double& x3, double& y3, double& x4, double& y4)
{

    // Esquina superior izquierda (x1, y1)
    x1 = x - (largo_efector / 2);
    y1 = y + (alto_efector / 2);

    // Esquina superior derecha (x2, y2)
    x2 = x + (largo_efector / 2);
    y2 = y + (alto_efector / 2);

    // Esquina inferior izquierda (x3, y3)
    x3 = x - (largo_efector / 2);
    y3 = y - (alto_efector / 2);

    // Esquina inferior derecha (x4, y4)
    x4 = x + (largo_efector / 2);
    y4 = y - (alto_efector / 2);
}

// CABLES
void calcular_cables(double x, double y, double largo_efector, double alto_efector, 
                     double largo_plano, double alto_plano,
                     double& L1, double& L2, double& q1, double& q2) 
{
    double x1, y1, x2, y2, x3, y3, x4, y4;
    calcular_esquinas(x, y, largo_efector, alto_efector, x1, y1, x2, y2, x3, y3, x4, y4);
    
    // Cable esquina superior izquierda M1 = (M1x, M1y)
    double M1x = 0;
    double M1y = alto_plano;

    // Cable esquina superior derecha M2 = (M2x, M2y)
    double M2x = largo_plano;
    double M2y = alto_plano;
    
    // LONGITUDES DE LOS CABLES
    L1 = std::sqrt(std::pow(x1 - M1x, 2) + std::pow(y1 - M1y, 2));
    L2 = std::sqrt(std::pow(x2 - M2x, 2) + std::pow(y2 - M2y, 2));
    
    // ÁNGULOS
    q1 = -std::atan((x1 - M1x) / (y1 - M1y)) * (180.0 / M_PI);
    q2 = std::atan((x2 - M2x) / (y2 - M2y)) * (180.0 / M_PI);
}

// FUNCIÓN PRINCIPAL MAIN
int 
main() 
{

    // PARÁMETROS
    double largo_plano = 100;
    double alto_plano = 100;
    double largo_efector = 10;
    double alto_efector = 10;
    double radio_rueda = 5;

    // SOLICITUD DE LAS COORDENADAS DEL EFECTOR FINAL
    std::cout << "\nCOORDENADAS INICIALES DEL EFECTOR FINAL\n";
    double x_inicial, y_inicial;
    std::cout << "Coordenada x inicial del efector final: ";
    std::cin >> x_inicial;
    std::cout << "Coordenada y inicial del efector final: ";
    std::cin >> y_inicial;
    std::cout << "\nCOORDENADAS FINALES DEL EFECTOR FINAL\n";
    double x_final, y_final;
    std::cout << "Coordenada x final del efector final: ";
    std::cin >> x_final;
    std::cout << "Coordenada y final del efector final: ";
    std::cin >> y_final;

    // VERIFICACIÓN DE LÍMITES
    std::string error_inicial = verificar_limites(x_inicial, y_inicial, largo_plano, alto_plano, largo_efector, alto_efector);
    if (!error_inicial.empty()) {
        std::cout << "\nERROR EN LA POSICIÓN INICIAL" << error_inicial << std::endl;
    }
    std::string error_final = verificar_limites(x_final, y_final, largo_plano, alto_plano, largo_efector, alto_efector);
    if (!error_final.empty()) {
        std::cout << "\nERROR EN LA POSICIÓN FINAL" << error_final << std::endl;
    }
    if (!error_inicial.empty() || !error_final.empty()) {
        return 1;
    }

    // PLANO
    plt::figure_size(1500, 1200);
    plt::xlim(-largo_plano * 0.25, largo_plano * 1.25);
    plt::ylim(-alto_plano * 0.5, alto_plano * 1.25);
    plt::xlabel("Eje X (centímetros)");
    plt::ylabel("Eje Y (centímetros)");
    plt::title("Robot por cables para el control de un efector final");

    // ESTRUCTURA
    std::vector<double> x_vertical1 = {0, 0};
    std::vector<double> y_vertical1 = {alto_plano, -(alto_plano / 2)};
    plt::plot(x_vertical1, y_vertical1, "k-");
    std::vector<double> x_vertical2 = {largo_plano, largo_plano};
    std::vector<double> y_vertical2 = {alto_plano, -(alto_plano / 2)};
    plt::plot(x_vertical2, y_vertical2, "k-");
    std::vector<double> x_horizontal1 = {0, largo_plano};
    std::vector<double> y_horizontal1 = {alto_plano, alto_plano};
    plt::plot(x_horizontal1, y_horizontal1, "k-");
    std::vector<double> x_horizontal2 = {0, largo_plano};
    std::vector<double> y_horizontal2 = {0, 0};
    plt::plot(x_horizontal2, y_horizontal2, "k-");
    std::vector<double> x_base1 = {-5, 5};
    std::vector<double> y_base1 = {-(alto_plano / 2), -(alto_plano / 2)};
    plt::plot(x_base1, y_base1, "k-");
    std::vector<double> x_base2 = {largo_plano - 5, largo_plano + 5};
    std::vector<double> y_base2 = {-(alto_plano / 2), -(alto_plano / 2)};
    plt::plot(x_base2, y_base2, "k-");

    // REPRESENTACIÓN RUEDAS (fijas)
    std::vector<double> x_wheels = {0.0, largo_plano};
    std::vector<double> y_wheels = {alto_plano, alto_plano};
    plt::scatter(x_wheels, y_wheels, radio_rueda);
    plt::scatter(x_wheels, y_wheels, radio_rueda/2);

    // ANIMACIÓN
    std::vector<double> trayectoria_x, trayectoria_y;
    
    for (int i = 0; i <= 100; ++i) {
        double t = i / 100.0;
        double x = x_inicial + (x_final - x_inicial) * t;
        double y = y_inicial + (y_final - y_inicial) * t;
        
        // ESQUINAS EFECTOR FINAL
        double x1, y1, x2, y2, x3, y3, x4, y4;
        calcular_esquinas(x, y, largo_efector, alto_efector, x1, y1, x2, y2, x3, y3, x4, y4);
        
        // REPRESENTACIÓN EFECTOR FINAL
        std::vector<double> xe = {x3, x4, x2, x1, x3};
        std::vector<double> ye = {y3, y4, y2, y1, y3};
        plt::plot(xe, ye, "k-");
        
        // CENTRO DEL EFECTOR
        std::vector<double> x_scatter = {x};
        std::vector<double> y_scatter = {y};
        plt::scatter(x_scatter, y_scatter, 2.0);
        
        // CABLES

        // Cable esquina superior izquierda M1 = (M1x, M1y)
        double M1x = 0;
        double M1y = alto_plano;
        std::vector<double> x_cable1 = {M1x, x1};
        std::vector<double> y_cable1 = {M1y, y1};
        plt::plot(x_cable1, y_cable1, "r-");

        // Cable esquina superior derecha M2 = (M2x, M2y)
        double M2x = largo_plano;
        double M2y = alto_plano;
        std::vector<double> x_cable2 = {M2x, x2};
        std::vector<double> y_cable2 = {M2y, y2};
        plt::plot(x_cable2, y_cable2, "r-");
        
        // TRAYECTORIA
        trayectoria_x.push_back(x);
        trayectoria_y.push_back(y);
        plt::plot(trayectoria_x, trayectoria_y, "b--");
        
        // FRAMES
        plt::pause(0.05);
        if (i < 100) {
            plt::clf();
            plt::xlim(-largo_plano * 0.25, largo_plano * 1.25);
            plt::ylim(-alto_plano * 0.5, alto_plano * 1.25);
            plt::xlabel("Eje X (centímetros)");
            plt::ylabel("Eje Y (centímetros)");
            plt::title("Robot por cables para el control de un efector final");
            plt::plot(x_vertical1, y_vertical1, "k-");
            plt::plot(x_vertical2, y_vertical2, "k-");
            plt::plot(x_horizontal1, y_horizontal1, "k-");
            plt::plot(x_horizontal2, y_horizontal2, "k-");
            plt::plot(x_base1, y_base1, "k-");
            plt::plot(x_base2, y_base2, "k-");
            plt::scatter(x_wheels, y_wheels, radio_rueda);
            plt::scatter(x_wheels, y_wheels, radio_rueda/2);
        }
    }

    // LONGITUD Y ÁNGULO DE CADA CABLE EN LA POSICIÓN INICIAL
    std::cout << "\nDATOS POSICIÓN INICIAL" << std::endl;
    double L1_inicial, L2_inicial, q1_inicial, q2_inicial;
    calcular_cables(x_inicial, y_inicial, largo_efector, alto_efector, largo_plano, alto_plano,
                   L1_inicial, L2_inicial, q1_inicial, q2_inicial);
    std::cout << "Longitud del cable L1 = " << L1_inicial << " cm" << std::endl;
    std::cout << "Longitud del cable L2 = " << L2_inicial << " cm" << std::endl;
    std::cout << "Ángulo del cable L1 (q1) = " << q1_inicial << " °" << std::endl;
    std::cout << "Ángulo del cable L2 (q2) = " << q2_inicial << " °" << std::endl;

    // LONGITUD Y ÁNGULO DE CADA CABLE EN LA POSICIÓN FINAL
    std::cout << "\nDATOS POSICIÓN FINAL" << std::endl;
    double L1_final, L2_final, q1_final, q2_final;
    calcular_cables(x_final, y_final, largo_efector, alto_efector, largo_plano, alto_plano,
                   L1_final, L2_final, q1_final, q2_final);
    std::cout << "Longitud del cable L1 = " << L1_final << " cm" << std::endl;
    std::cout << "Longitud del cable L2 = " << L2_final << " cm" << std::endl;
    std::cout << "Ángulo del cable L1 (q1) = " << q1_final << " °" << std::endl;
    std::cout << "Ángulo del cable L2 (q2) = " << q2_final << " °" << std::endl;

    // LONGITUD DE CABLE ELONGADA / RECOGIDA Y ÁNGULO DE GIRO DE CADA POLEA
    std::cout << "\nLONGITUD DE CABLE ELONGADA / RECOGIDA Y ÁNGULO DE GIRO DE CADA POLEA" << std::endl;
    double L1_movido = L1_final - L1_inicial;
    double L2_movido = L2_final - L2_inicial;
    double P1_movido_radianes = L1_movido / radio_rueda;
    double P2_movido_radianes = L2_movido / radio_rueda;
    std::cout << "Longitud de cable elongada / recogida por el cable L1 = " << L1_movido << " cm" << std::endl;
    std::cout << "Longitud de cable elongada / recogida por el cable L2 = " << L2_movido << " cm" << std::endl;
    std::cout << "Ángulo girado por la polea P1 = " << P1_movido_radianes << " radianes" << std::endl;
    std::cout << "Ángulo girado por la polea P1 = " << (P1_movido_radianes * 180.0 / M_PI) << " °" << std::endl;
    std::cout << "Ángulo girado por la polea P2 = " << P2_movido_radianes << " radianes" << std::endl;
    std::cout << "Ángulo girado por la polea P2 = " << (P2_movido_radianes * 180.0 / M_PI) << " °" << std::endl;

    plt::show();
    return 0;
}

File: scripts/python/cinematic_model_v2.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
from matplotlib.path import Path
import matplotlib.patches as patches

# PARAMETERS
largo_plano = 100
alto_plano = 100
largo_efector = 10
alto_efector = 10
radio_rueda = 5

# REQUEST FOR THE COORDINATES OF THE END-EFFECTOR
print("\nCOORDENADAS INICIALES DEL EFECTOR FINAL")
x_inicial = float(input('Coordenada x inicial del efector final: '))
y_inicial = float(input('Coordenada y inicial del efector final: '))
print("\nCOORDENADAS FINALES DEL EFECTOR FINAL")
x_final = float(input('Coordenada x final del efector final: '))
y_final = float(input('Coordenada y final del efector final: '))

# LIMIT VERIFICATION
def verificar_limites(x, y):
    error_msg = ""
    if x >= (largo_plano - (largo_efector / 2)) or x <= (largo_efector / 2):
        error_msg += f"\nFUERA DEL LÍMITE en x = {x}"
    if y >= (alto_plano - (alto_efector / 2)) or y <= (alto_efector / 2):
        error_msg += f"\nFUERA DEL LÍMITE en y = {y}"
    return error_msg
error_inicial = verificar_limites(x_inicial, y_inicial)
if error_inicial:
    print("\nERROR EN LA POSICIÓN INICIAL" + error_inicial)
error_final = verificar_limites(x_final, y_final)
if error_final:
    print("\nERROR EN LA POSICIÓN FINAL" + error_final)
if error_inicial or error_final:
    exit()

# PLANE
fig, ax = plt.subplots(figsize=(20, 16))
ax.set_xlim([-largo_plano * 0.25, largo_plano * 1.25])
ax.set_ylim([-alto_plano * 0.5, alto_plano * 1.25])
ax.set_xlabel('Eje X (centímetros)')
ax.set_ylabel('Eje Y (centímetros)')
ax.set_title('Robot por cables para el control de un efector final')

# STRUCTURE
ax.plot([0, 0], [alto_plano, -(alto_plano / 2)], 'k', linewidth=5)
ax.plot([largo_plano, largo_plano], [alto_plano, -(alto_plano / 2)], 'k', linewidth=5)
ax.plot([0, largo_plano], [alto_plano, alto_plano], 'k', linewidth=5)
ax.plot([0, largo_plano], [0, 0], 'k', linewidth=5)
ax.plot([-5, 5], [-(alto_plano / 2), -(alto_plano / 2)], 'k', linewidth=5)
ax.plot([largo_plano - 5, largo_plano + 5], [-(alto_plano / 2), -(alto_plano / 2)], 'k', linewidth=5)

# PULLEYS REPRESENTATION
ax.plot(0, alto_plano, 'ko', markersize=radio_rueda, markerfacecolor='black')
ax.plot(0, alto_plano, 'wo', markersize=radio_rueda / 2, markerfacecolor='black')
ax.plot(largo_plano, alto_plano, 'ko', markersize=radio_rueda, markerfacecolor='black')
ax.plot(largo_plano, alto_plano, 'wo', markersize=radio_rueda / 2, markerfacecolor='black')

# UPDATE OF THE POSITION OF THE DISPLACED ELEMENTS
efector, = ax.plot([], [], 'black', linewidth=2)
centro_efector, = ax.plot([], [], 'ko', markersize=2, markerfacecolor='black')
cable1, = ax.plot([], [], 'r', linewidth=2)
cable2, = ax.plot([], [], 'r', linewidth=2)
trayectoria, = ax.plot([], [], 'b--', linewidth=1, alpha=0.5)

# STORAGE OF POINTS OF THE TRAJECTORY TAKEN
trayectoria_x = []
trayectoria_y = []

# END-EFFECTOR
def calcular_esquinas(x, y):
    # Top left corner (x1, y1)
    x1 = x - (largo_efector / 2)
    y1 = y + (alto_efector / 2)
    # Top right corner (x2, y2)
    x2 = x + (largo_efector / 2)
    y2 = y + (alto_efector / 2)
    # Bottom left corner (x3, y3)
    x3 = x - (largo_efector / 2)
    y3 = y - (alto_efector / 2)
    # Bottom right corner (x4, y4)
    x4 = x + (largo_efector / 2)
    y4 = y - (alto_efector / 2)
    return x1, y1, x2, y2, x3, y3, x4, y4

# CABLES
def calcular_cables(x, y):
    x1, y1, x2, y2, _, _, _, _ = calcular_esquinas(x, y)
    # Cable top left corner M1 = (M1x, M1y)
    M1x = 0
    M1y = alto_plano
    # Cable top right corner M2 = (M2x, M2y)
    M2x = largo_plano
    M2y = alto_plano
    # CABLE LENGTHS
    L1 = np.sqrt((x1 - M1x) ** 2 + (y1 - M1y) ** 2)
    L2 = np.sqrt((x2 - M2x) ** 2 + (y2 - M2y) ** 2)
    # ANGLES
    q1 = -np.degrees(np.arctan((x1 - M1x) / (y1 - M1y)))
    q2 = np.degrees(np.arctan((x2 - M2x) / (y2 - M2y)))
    return L1, L2, q1, q2

# ANIMATION DATA INITIALIZATION
def init():
    efector.set_data([], [])
    centro_efector.set_data([], [])
    cable1.set_data([], [])
    cable2.set_data([], [])
    trayectoria.set_data([], [])
    return efector, centro_efector, cable1, cable2, trayectoria

# ANIMATION
def animate(i):
    
    # FRAMES
    t = i / 100
    x = x_inicial + (x_final - x_inicial) * t
    y = y_inicial + (y_final - y_inicial) * t
    
    # END-EFFECTOR CORNERS
    x1, y1, x2, y2, x3, y3, x4, y4 = calcular_esquinas(x, y)
    
    # END-EFFECTOR REPRESENTATION
    xe = [x3, x4, x2, x1, x3]
    ye = [y3, y4, y2, y1, y3]
    efector.set_data(xe, ye)
    centro_efector.set_data(x, y)

    # CABLES
    # Cable top left corner M1 = (M1x, M1y)
    M1x = 0
    M1y = alto_plano
    # Cable top right corner M2 = (M2x, M2y)
    M2x = largo_plano
    M2y = alto_plano

    # UPDATE CABLE POSITION
    cable1.set_data([M1x, x1], [M1y, y1])
    cable2.set_data([M2x, x2], [M2y, y2])
    
    # TRAJECTORY UPDATE
    trayectoria_x.append(x)
    trayectoria_y.append(y)
    trayectoria.set_data(trayectoria_x, trayectoria_y)
    return efector, centro_efector, cable1, cable2, trayectoria

# ANIMATION CREATION
ani = FuncAnimation(fig, 
                    animate, 
                    frames=100, 
                    init_func=init,
                    blit=True, 
                    interval=50, 
                    repeat=False)

# LENGTH AND ANGLE OF EACH CABLE IN THE INITIAL POSITION
print("\nDATOS POSICIÓN INICIAL")
L1_inicial, L2_inicial, q1_inicial, q2_inicial = calcular_cables(x_inicial, y_inicial)
print("Longitud del cable L1 =", L1_inicial, "cm")
print("Longitud del cable L2 =", L2_inicial, "cm")
print("Ángulo del cable L1 (q1) =", q1_inicial, "°")
print("Ángulo del cable L2 (q2) =", q2_inicial, "°")

# LENGTH AND ANGLE OF EACH CABLE IN THE FINAL POSITION
print("\nDATOS POSICIÓN FINAL")
L1_final, L2_final, q1_final, q2_final = calcular_cables(x_final, y_final)
print("Longitud del cable L1 =", L1_final, "cm")
print("Longitud del cable L2 =", L2_final, "cm")
print("Ángulo del cable L1 (q1) =", q1_final, "°")
print("Ángulo del cable L2 (q2) =", q2_final, "°")

# CABLE LENGTH EXTENDED / RETRACTED AND TURNING ANGLE OF EACH PULLEY
print("\nLONGITUD DE CABLE ELONGADA / RECOGIDA Y ÁNGULO DE GIRO DE CADA POLEA")
L1_movido = L1_final - L1_inicial
L2_movido = L2_final - L2_inicial
P1_movido_radianes = L1_movido / radio_rueda
P2_movido_radianes = L2_movido / radio_rueda
print("Longitud de cable elongada / recogida por el cable L1 =", L1_movido, "cm")
print("Longitud de cable elongada / recogida por el cable L2 =", L2_movido, "cm")
print("Ángulo girado por la polea P1 =", P1_movido_radianes, "radianes")
print("Ángulo girado por la polea P1 =", np.degrees(P1_movido_radianes), "°")
print("Ángulo girado por la polea P2 =", P2_movido_radianes, "radianes")
print("Ángulo girado por la polea P2 =", np.degrees(P2_movido_radianes), "°")

plt.show()
